buildSrc/src/main/kotlin/pklJavaLibrary.gradle.kts

```
@@ -80,9 +110,60 @@ val workAroundKotlinGradlePluginBug by
     }
   }
 
+val truffleJavacArgs =
+  listOf(
+    // TODO: determine correct limits for Truffle specializations
+    // (see https://graalvm.slack.com/archives/CNQSB2DHD/p1712380902746829)
+    "-Atruffle.dsl.SuppressWarnings=truffle-limit"
+  )
+
+val jpmsJavacArgs =
+  listOf(
+    // Required for use of `sun.misc.Unsafe` in Truffle.
+    "--add-modules=jdk.unsupported"
+  )
+
+val javacArgsProvider = CommandLineArgumentProvider { jpmsJavacArgs }
+
 tasks.compileJava {
+  javaCompiler = info.javaCompiler
   dependsOn(workAroundKotlinGradlePluginBug)
-  // TODO: determine correct limits for Truffle specializations
-  // (see https://graalvm.slack.com/archives/CNQSB2DHD/p1712380902746829)
-  options.compilerArgs.add("-Atruffle.dsl.SuppressWarnings=truffle-limit")
+  options.compilerArgumentProviders.add(CommandLineArgumentProvider { truffleJavacArgs })
 }
+
+tasks.withType<JavaCompile>().configureEach {
+  val jvmTarget = JavaVersion.toVersion(info.jvmTarget)
+  javaCompiler = info.javaCompiler
+  sourceCompatibility = jvmTarget.majorVersion
+  targetCompatibility = jvmTarget.majorVersion
+
+  options.compilerArgumentProviders.add(javacArgsProvider)
+}
+
+tasks.withType<JavaExec>().configureEach { jvmArgs(jpmsJavacArgs) }
+
+fun Test.configureJdkTestTask(launcher: Provider<JavaLauncher>) {
+  useJUnitPlatform()
+  javaLauncher = launcher
+  jvmArgumentProviders.add(javacArgsProvider)
+  jvmArgumentProviders.add(
+    CommandLineArgumentProvider { testProperties.map { (key, value) -> "-D$key=$value" } }
+  )
```

[nit] don't need to use a provider since none of the argument values need to stay lazy.

```suggestion
  jvmArgs.addAll(jpmsJavacArgs)
  jvmArgs.addAll(testProperties.map { (key, value) -> "-D$key=$value" })
```